File: test/Dapper.FSharp.MySQL.Tests/SetUpDbContainer.fs
namespace Dapper.FSharp.MySQL.Tests

open System
open System.IO
open NUnit.Framework
open Testcontainers.MySql

[<SetUpFixture>]
type SetUpDbContainer() =
    let mySqlContainer () =
        task {
            let uri = DirectoryInfo(Path.Combine(__SOURCE_DIRECTORY__, ""))
            let fi = FileInfo(Path.Combine(uri.FullName, "init.sql"))
            let container =
                MySqlBuilder()
                    .WithResourceMapping(fi, "/docker-entrypoint-initdb.d/")
                    .WithDatabase("fsharp_dapper_test_db")
                    .WithUsername("admin")
                    .WithPassword("admin")
                    .Build()

            do TestContext.Progress.WriteLine($"[{DateTime.Now:``HH:mm:ss:fff``}] MySqlTests: Starting MySqlContainer...")
            do! container.StartAsync().ConfigureAwait(false)
            do TestContext.Progress.WriteLine($"[{DateTime.Now:``HH:mm:ss:fff``}] MySqlTests: MySqlContainer has been started.")
            return container
        }

    static member val MySqlContainer : MySqlContainer = Unchecked.defaultof<_> with get, set

    [<OneTimeSetUp>]
    member _.``create MySqlContainer`` () =
        task {
            Dapper.FSharp.MySQL.OptionTypes.register()
            let! mySqlContainer = mySqlContainer ()
            SetUpDbContainer.MySqlContainer <- mySqlContainer
        }

    [<OneTimeTearDown>]
    member _.``dispose MySqlContainer`` () =
        task {
            match SetUpDbContainer.MySqlContainer with
            | null -> return ()
            | _ ->
                do!
                    (SetUpDbContainer.MySqlContainer :> IAsyncDisposable).DisposeAsync()
                do TestContext.Progress.WriteLine($"[{DateTime.Now:``HH:mm:ss:fff``}] MySqlTests: MySqlContainer has been stopped.")
        }

File: test/CancellationTests/SetUpDbContainer.cs
using Testcontainers.MySql;

namespace TestProject1;

[SetUpFixture]
public class SetUpDbContainer
{
    public static MySqlContainer? MySqlContainer { get; private set; }

    [OneTimeSetUp]
    public async Task CreateMySqlContainer ()
    {
        var mySqlContainer = await UpMySqlContainer();
        SetUpDbContainer.MySqlContainer = mySqlContainer;
    }

    [OneTimeTearDown]
    public async Task DisposeMySqlContainer()
    {
        if (SetUpDbContainer.MySqlContainer is not null)
        {
            await SetUpDbContainer.MySqlContainer.DisposeAsync();
        }
    }

    private async Task<MySqlContainer> UpMySqlContainer()
    {
        var container =
            new MySqlBuilder()
                .WithImage("mysql:8.0.33")
                .WithDatabase("fsharp_dapper_test_db")
                .WithUsername("admin")
                .WithPassword("admin")
                .Build();

        TestContext.Progress.WriteLine($"[{DateTime.Now:``HH:mm:ss:fff``}] MySqlTests: Starting MySqlContainer...");
        await container.StartAsync().ConfigureAwait(false);
        TestContext.Progress.WriteLine($"[{DateTime.Now:``HH:mm:ss:fff``}] MySqlTests: MySqlContainer has been started.");
        return container;
    }
}
